mempool/src/pool.rs

```diff
@@ -0,0 +1,330 @@
+use std::cmp::Ord;
+use std::cmp::Ordering;
+use std::collections::BTreeSet;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fmt::Debug;
+use std::rc::Rc;
+
+use parity_scale_codec::Encode;
+use thiserror::Error;
+
+use common::chain::transaction::Transaction;
+use common::chain::transaction::TxInput;
+use common::chain::OutPoint;
+use common::primitives::amount::Amount;
+use common::primitives::Id;
+use common::primitives::Idable;
+use common::primitives::H256;
+
+// TODO this willbe defined elsewhere (some of limits.rs file)
+const MAX_BLOCK_SIZE_BYTES: usize = 1_000_000;
+
+const MEMPOOL_MAX_TXS: usize = 1_000_000;
```

Comment:
This certainly has to be managed in a much better way. The mempool should depend on how much memory the node wants to consume, not an arbitrary constant. Again, given that none of this is final, maybe this is OK for now. In bitcoin core there's a cache size that determines when the mempool is full.

Reply:
Agree completely. Is memory consumption the only factor in determining whether the mempool is full? Number of TXs is irrelevant?

Reply:
I can't think of a reason why the number of txs would be relevant. Who cares if you have 10000 or 1000000 txs in your mempool?

Reply:
Yeah I don't know what happened to my comment from last week, I guess someone resolved it. This will get massive, insanely massive

Reply:
We will need a way to estimate the current size of the mempool. The trait GetMemoryUsage (currently only implemented by mocks) is a preparation for this.

Reply:
I think a good start would be the total serialized size of all transactions in it.

About the constant up there, we have now configuration for different subsystems. The mempool is not a subsystem right now, but has to become one at some point. Do we do that in this PR? I don't know. Up to you. But I think creating a subsystem out of it is not a very big deal. All you have to do is create an interface (see how p2p and chainstate do it), and initialize p2p in the node crate. Then, you'll be able to pass a new configuration object, in which such numbers are available. This will also help us understand (during the review) how the mempool works from the outside since the interface will be exposed. I think that's a good idea to pursue.